chia/cmds/check_wallet_db.py

```diff
@@ -0,0 +1,340 @@
+                missing_wallet_ids.append(d.wallet_id)
+            elif d.wallet_type != wallet_id_to_type[d.wallet_id]:
+                wrong_type[(d.hardened, d.wallet_id, d.wallet_type, wallet_id_to_type[d.wallet_id])].append(
+                    d.derivation_index
+                )
+
+        if len(invalid_wallet_types) > 0:
+            errors.append(f"Invalid wallet_types in derivation_paths table: {invalid_wallet_types}")
+
+        if len(missing_wallet_ids) > 0:
+            errors.append(
+                f"Wallet IDs found in derivation_paths table, but not in users_wallets table: {missing_wallet_ids}"
+            )
+
+        for k, v in wrong_type.items():
+            errors.append(
+                f"""{["  ", "un"][int(k[0])]}hardened Wallet ID {k[1]} uses type {k[2]} in derivation_paths, """
+                f"""but type {k[3]} in wallet table at these derivation indices: {v}"""
+            )
+
+        return errors
+
+    async def scan(self, db_path: Path) -> None:
+        self.db_wrapper = await DBWrapper2.create(
+            database=db_path,
+            reader_count=self.config.get("db_readers", 4),
+            log_path=self.sql_log_path,
+            synchronous=db_synchronous_on("auto"),
+        )
+        # TODO: Pass down db_wrapper
+        wallets = await self.get_all_wallets()
+        derivation_paths = await self.get_derivation_paths()
+        try:
+            errors = []
+
+            if self.verbose:
+                await self.show_tables()
+                print_min_max_derivation_for_wallets(derivation_paths)
+
+            errors.extend(await self.check_wallets())
+            errors.extend(self.check_wallets_missing_derivations(wallets, derivation_paths))
+            errors.extend(self.check_unexpected_derivation_entries(wallets, derivation_paths))
+            errors.extend(self.check_derivations_are_compact(wallets, derivation_paths))
+            if len(errors) > 0:
+                print("\n    ---- Errors Found ----")
+                print("\n".join(errors))
+                sys.exit(2)
+            else:
+                print("No errors found.\n")
+        finally:
+            await self.db_wrapper.close()
+
+
+async def scan(root_path: str, db_path: Optional[str] = None, *, verbose: bool = False) -> None:
+
+    if db_path is None:
+        wallet_db_path = Path(root_path) / "wallet" / "db"
+        wallet_db_paths = list(wallet_db_path.glob("blockchain_wallet_*.sqlite"))
+    else:
+        wallet_db_paths = [Path(db_path)]
+
+    for wallet_db_path in wallet_db_paths:
+        w = WalletDBReader()
+        w.verbose = verbose
+        print(f"Reading {wallet_db_path}")
+        await w.scan(Path(wallet_db_path))
+
+
+if __name__ == "__main__":
+    loop = asyncio.get_event_loop()
+    loop.run_until_complete(scan("", sys.argv[1]))
+
+
+def test_check_for_gaps_end_lt_start() -> None:
+    with pytest.raises(ValueError):
+        _ = check_for_gaps([], 2, 0)
```

Comment:
it seems a bit odd to put test functions in here

Reply:
Yes - Just pushing this up so people can use it. Still need to move some things around, and add some tests.